### PR TITLE
feat: add banner support for communities

### LINKED ISSUE #5403: Add banner support to backend (status-go)

### PR DESCRIPTION
**Pull request by stefandunca**
Add banner image as a special "banner" `IdentityImage` beside "thumbnail" and "large"

Changes to `images` module

- Refactor `EncodeToBestSize` as `EncodeToLimits` to accept arbitrary dimmensions to allow for custom size
- Define `DimensionLimits` for banner not to exceed 450 KB

Closes [status-desktop #5403](https://github.com/status-im/status-desktop/issues/5403)

### CODE REVIEW

File: images/encode.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ func renderJpeg(w io.Writer, m image.Image, config EncodeConfig) error {
 	return jpeg.Encode(w, m, o)
 }
 
-func EncodeToBestSize(bb *bytes.Buffer, img image.Image, size ResizeDimension) error {
+func EncodeToLimits(bb *bytes.Buffer, img image.Image, bounds DimensionLimits) error {
```
Comment (Samyoul) [inline images/encode.go:28]: Nice refactor

**Review by richard-ramos — APPROVED** [review]

File: images/meta.go [hunk 2]
```diff
@@ -58,3 +60,12 @@ type DimensionLimits struct {
 
 type ImageType uint
 type ResizeDimension uint
+
+const BannerIdentityName string = "banner"
```
Comment (Samyoul) [inline images/meta.go]: Maybe you could move this up to line 24 to be with the other image names

File: images/meta.go [hunk 3]
```diff
@@ -58,3 +60,12 @@ type DimensionLimits struct {
...
+
+func GetBannerDimension() DimensionLimits {
+	return DimensionLimits{
+		Ideal: 460800, // We want the best quality supported
```
Comment (Samyoul) [inline images/meta.go]: Why do we want to aim for the maximum size? Shouldn't at least try to save space on the image?

File: protocol/requests/cropped_image.go [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package requests
+
+type CroppedImage struct {
+	ImagePath string `json:"imagePath"`
+	X         int    `json:"x"`
+	Y         int    `json:"y"`
+	Width     int    `json:"width"`
+	Height    int    `json:"height"`
+}
```
Comment (Samyoul) [inline protocol/requests/cropped_image.go]: This type should probably live in the `images` package

**Review by Samyoul — COMMENTED** [review]
Looks good. A few points nothing major. One thing that I don't see is any tests for the banner functionality, do you feel confident in adding these?

File: images/meta.go (hunk 3 above)
Comment (stefandunca) [reply images/meta.go]: Thanks for the heads up. I now realize I didn't give much thought to the importance of size to traffic and storage. I only take into consideration that
- community admins would want maximum quality
- a user won't be part of many communities and sync many logos
- the design requirement is that we are not allowed to have images bigger than 450 KB

Should we conduct an analysis similar to the logo image [here](https://github.com/status-im/status/issues/36#issuecomment-694970473) ?

Is it enough for now to use a rough estimation and test the effects? I changed it to `102400` which looks good enough in my tests

**Review by stefandunca — COMMENTED** [review]
> ... I don't see is any tests for the banner functionality, do you feel confident in adding these?

@Samyoul, how about I add integration tests to check that banner image processing respects the imposed size limits? However, that means working with big images, which might make tests longer than a few milliseconds. Is this a problem?

Do you have anything else in mind for the test scope?

Also, I'm not familiar with image processing in go. Any hints on what should I use to generate input test images so that I don't pollute the test source code with big files?

File: images/meta.go (hunk 3 above)
Comment (Samyoul) [reply images/meta.go]: Yes, that sounds fine to me. Yeah bandwidth is always a big concern with the nature of our application

**Comment by Samyoul** [comment]
Perhaps check the `IdentityImage` components of `ToCommunityDescription()`

**Comment by stefandunca** [comment]
@Samyoul, I did all the requested changes, so please go ahead and have another look.

Also, after some tests with different compression quality, base resolution, and image complexity and talking to @John-44 I raised the ideal image quality to 300KB, which is an edge case, to target maximum quality.

I also changed the behaviour for banners not to upscale to target quality but only shrink if necessary. We can handle the upscaling in the UIs and save some space/traffic.

**Review by Samyoul — APPROVED** [review]
Great, looks good

**Pull request merged**